### PR TITLE
Enforce code style analysis shipped with .NET SDK

### PR DESCRIPTION
**Pull request by huoyaoyuan**
The same thing of https://github.com/ppy/osu-framework/pull/5121

### CODE REVIEW

**Comment by huoyaoyuan** [comment]
If android build is complaining about compatibility, we can wait for the .NET 6 migration to be merged first.

File: osu.Game.Tests/Visual/Collections/TestSceneManageCollectionsDialog.cs [hunk 1]
```diff
@@ -141,7 +141,8 @@ public void TestCollectionNameCollisions()
         {
             AddStep("add dropdown", () =>
             {
-                Add(new CollectionFilterDropdown
+                Add(
+                    new CollectionFilterDropdown
```
Comment (bdach) [inline osu.Game.Tests/Visual/Collections/TestSceneManageCollectionsDialog.cs:145]: Dislike this change, old code was preferable.

**Review by bdach — COMMENTED** [review]
Seems mostly OK

File: osu.Game.Tests/Visual/Collections/TestSceneManageCollectionsDialog.cs (hunk 1 above)
Comment (huoyaoyuan) [reply osu.Game.Tests/Visual/Collections/TestSceneManageCollectionsDialog.cs:145]: This is tricky to satisfy both analyzers.

Comment (bdach) [reply osu.Game.Tests/Visual/Collections/TestSceneManageCollectionsDialog.cs:145]: Something like this is idiomatic in the codebase and doesn't seem to trigger the .net analyser:

```diff
diff --git a/osu.Game.Tests/Visual/Collections/TestSceneManageCollectionsDialog.cs b/osu.Game.Tests/Visual/Collections/TestSceneManageCollectionsDialog.cs
index d5983ac827..888002eb36 100644
--- a/osu.Game.Tests/Visual/Collections/TestSceneManageCollectionsDialog.cs
+++ b/osu.Game.Tests/Visual/Collections/TestSceneManageCollectionsDialog.cs
@@ -141,15 +141,13 @@ public void TestCollectionNameCollisions()
         {
             AddStep("add dropdown", () =>
             {
-                Add(
-                    new CollectionFilterDropdown
-                    {
-                        Anchor = Anchor.TopRight,
-                        Origin = Anchor.TopRight,
-                        RelativeSizeAxes = Axes.X,
-                        Width = 0.4f,
-                    }
-                );
+                Add(new CollectionFilterDropdown
+                {
+                    Anchor = Anchor.TopRight,
+                    Origin = Anchor.TopRight,
+                    RelativeSizeAxes = Axes.X,
+                    Width = 0.4f,
+                });
             });
             AddStep("add two collections with same name", () => manager.Collections.AddRange(new[]
             {
```

Comment (huoyaoyuan) [reply osu.Game.Tests/Visual/Collections/TestSceneManageCollectionsDialog.cs:145]: I remember that resharper requires 4 chars of indentation of the braces here, while VS requires no indentation.

Comment (bdach) [reply osu.Game.Tests/Visual/Collections/TestSceneManageCollectionsDialog.cs:145]: Yes, and I'm fixing that conflict by merging closing curly and round brackets into a single line. Which makes the analysers agree again.

**Review by bdach — DISMISSED** [review]
seems okay

**Review by bdach — DISMISSED** [review]

File: Directory.Build.props [hunk 2]
```diff
@@ -20,6 +20,7 @@
     <AdditionalFiles Include="$(MSBuildThisFileDirectory)CodeAnalysis\BannedSymbols.txt" />
   </ItemGroup>
   <PropertyGroup Label="Code Analysis">
+    <EnforceCodeStyleInBuild Condition="'$(EnforceCodeStyleInBuild)' == ''">true</EnforceCodeStyleInBuild>
```
Comment (peppy) [inline Directory.Build.props:23]: I am not sure we can warrant this being always on.

```md

## master

Clean: 14.32s 13.90s
No change: 2.93s 3.01s
Change: 10.66s 10.79s

Desktop no change: 1.64s 1.66s
Desktop change: 7.2s 7.47s

## this pr

Clean: 19.43s 17.32s
No change: 2.97s 3.07s
Change: 13.98s 14.13s

Desktop no change: 1.62s 1.63s
Desktop change: 9.06s 9.18s
```

For full project builds, that's 30% overhead. For desktop builds, 24%. I don't want this kind of overhead on my builds. So if we're implementing this it needs to either be off by default, or there needs to be an easy way to turn it off.

**Review by peppy — CHANGES_REQUESTED** [review]
Actual code changes look fine, but..

File: Directory.Build.props (hunk 2 above)
Comment (bdach) [reply Directory.Build.props:23]: For the record I'm fine with only enabling this on CI. I'd like to think it shouldn't be actually needed very much on desktop, as any decent IDE (*side-eyes vs code*) should be able to display the inspections in real time as errors.

File: .editorconfig [hunk 3]
```diff
@@ -168,28 +187,15 @@ dotnet_style_prefer_is_null_check_over_reference_equality_method = true:silent
 #Style - C# 8 features
 csharp_prefer_static_local_function = true:warning
 csharp_prefer_simple_using_statement = true:silent
-csharp_style_prefer_index_operator = false:silent
-csharp_style_prefer_range_operator = false:silent
+csharp_style_prefer_index_operator = true:warning
+csharp_style_prefer_range_operator = true:warning
```
Comment (PercyDan54) [inline .editorconfig:191]: Just asking why this is changed [over](https://github.com/ppy/osu/commit/c457571da6d69f41ac7722bda6532fa99722c611) and [over](https://github.com/ppy/osu/pull/13789/files)? So which one is the preferred way of this

Comment (bdach) [reply .editorconfig:191]: This is a good point, there have been instances of us asking to revert range syntax usage due to it being unreadable. Probably best to revert.

Comment (huoyaoyuan) [reply .editorconfig:191]: I copied the whole editorconfig from framework repo, to keep consistency.

The severity in editorconfig only affects VS IDE, not command line building. But generally if it's not enforced, setting to slient or suggestion is better.

File: Directory.Build.props (hunk 2 above)
Comment (huoyaoyuan) [reply Directory.Build.props:23]: VS will skip analyzers when hitting F5 to debug, but other IDEs can be different.
I didn't get a measurable difference at framework repo, but if there is as you shown, we can just enable it in CI, and probably put under code quality leg.

**Review by peppy — APPROVED** [review]

**Review by frenzibyte — APPROVED** [review]

**Pull request merged**